doc/03_research/03_planning/06_state_machine_design.md

```diff
@@ -0,0 +1,278 @@
+
+Transition:
+
+* From `DECIDE_LANE_CHANGE` by `RoadOption.CHANGELANELEFT` or `obstacle_in_lane` or `leader_vehicle_speed < LEADERTHRESHOLD`
+
+This state performs a lane change to the lane on the left.
+
+If the lane is for oncoming traffic, it should be checked if there is enough time to complete two lane changes plus a buffer for the length of an obstacle without an accident.
+
+If the lane is for traffic flowing in the same direction, It should be checked if the lane is free and if there is no faster moving vehicles coming from behind.
+
+If an obstacle or a slow leading vehicle are the reasons for the lane change, to join the same lane after the obstacle is passed, the state changes to `CHANGE_LANE_RIGHT`.
+
+### CHANGE_LANE_RIGHT
+
+Transition:
+
+* From `DECIDE_LANE_CHANGE` by `RoadOption.CHANGELANERIGHT` or `emergency_vehicle_in_front`
+* From `CHANGE_LANE_LEFT` by `passing_obstacle` or `slow_leading_vehicle`
+
+For changing to the right lane it is assumed, that the traffic in this lane flows in the driving direction of the ego vehicle.
+
+The lane change should be performed if the lane is free and there are no fast moving cars behind us that would crash with the ego vehicle.
+
+## Intersection state machine
+
+![img.png](../../00_assets/Intersection_SM.png)
+
+Transition:
+
+* From `driving state machine` by `intersection_detected`
+
+This state machine handles the passing of an intersection.
+
+### APPROACH_INTERSECTION
+
+Slow down in the approach of an intersection.
+
+If the intersection is `unsignalized` and `clear` change to the `ENTER_INTERSECTION` state.
+
+If there are is a traffic light or a stop sign at the intersection change to the stop_sign/traffic state machine.
+
+### IN_INTERSECTION
+
+Transition:
+
+* From `STOP_SIGN/TRAFFIC SM` by `clearing the traffic light, stop sign`
+* From `APPROACH_INTERSECTION` by `detecting an unsignalized and cleared intersection`
+
+After the approach of the intersection and clear a possible traffic light/stop sign, the ego vehicle enters the intersection.
+
+From there the RoadOption decides in which direction the ego vehicle should turn.
+
+### TURN_LEFT
+
+Transition:
+
+* From `IN_INTERSECTION` by `RoadOption.LEFT`
+
+Check for pedestrians on the driving path. If the path is clear of pedestrians, make sure there will be no crashes during the turning process with oncoming traffic.
+
+### STRAIGHT
+
+Transition:
+
+* From `IN_INTERSECTION` by `RoadOption.STRAIGHT`
+
+Check if there is a vehicle running a red light in the intersection. Pass the intersection.
+
+### TURN_RIGHT
+
+Transition:
+
+* From `IN_INTERSECTION` by `RoadOption.RIGHT`
+
+Check for pedestrians on the driving path. If the path is clear of pedestrians, make sure there will be no crashes during the turning process with crossing traffic.
+
+### LEAVE_INTERSECTION
+
+Transition:
+
+* From `TURN_RIGHT`, `STRAIGHT` or `TURN_LEFT` by passing a distance from the intersection.
+
+## Stop sign/traffic light state machine
```

Comment:
Vermutlich ist es nicht ratsam, bei Gelb grundsätzlich zu versuchen, noch vor der Kreuzung stehenzubleiben. Das provoziert Auffahrunfälle. Im ungünstigsten Fall bleiben wir erst in der Kreuzung stehen, weil wir nicht schnell genug bremsen können, und warten dann in der Kreuzung bis es Grün wird.

Reply:
Im ungünstigen Fall fahren wir dann über eine rote Ampel und kassieren einen Penalty, soweit ich weiß, gibt es aber keinen Penalty dafür, dass ein Auto in uns reinfährt. Dies wurde eher aus Sicherheitsgründen eingefügt.

Reply:
Ok, das macht Sinn, wobei ich mir nicht sicher bin, ob Carla einen Auffahrunfall nicht einfach wie eine Kollision behandelt.